### PR TITLE
Slug

### PR DESCRIPTION
**Pull request by hadalin**

### CODE REVIEW

File: resolwe/flow/serializers/base.py [hunk 1]
```diff
@@ -56,3 +59,10 @@ def __init__(self, instance=None, data=empty, **kwargs):
     def request(self):
         """Extract request object from serializer context."""
         return self.context.get('request', None)
+
+    def save(self):
+        """Override save method to catch handled errors and repackage them as 400 errors."""
+        try:
+            super(ResolweBaseSerializer, self).save()
+        except DatabaseError as database_error:
```
Comment (hadalin) [inline resolwe/flow/serializers/base.py]: Not the best idea.

A better solution would be to raise a specific resolwe error [here](https://github.com/genialis/resolwe/blob/7b8ee3e4de21affa7cf18d73e9f0acb5556c06ec/resolwe/flow/models/fields.py#L178) and catch it here. That would entail an `excpetions` package with that error defined.

File: resolwe/flow/serializers/base.py [hunk 2]
```diff
@@ -56,3 +59,10 @@ def __init__(self, instance=None, data=empty, **kwargs):
     def request(self):
         """Extract request object from serializer context."""
         return self.context.get('request', None)
+
+    def save(self):
+        """Override save method to catch handled errors and repackage them as 400 errors."""
+        try:
+            return super(ResolweBaseSerializer, self).save()
```
Comment (dblenkus) [inline resolwe/flow/serializers/base.py]: `return super().save(**kwargs)`

File: resolwe/flow/serializers/base.py [hunk 3]
```diff
@@ -56,3 +59,10 @@ def __init__(self, instance=None, data=empty, **kwargs):
     def request(self):
         """Extract request object from serializer context."""
         return self.context.get('request', None)
+
+    def save(self):
```
Comment (dblenkus) [inline resolwe/flow/serializers/base.py]: `def save(self, **kwargs):`

File: resolwe/flow/serializers/base.py (hunk 1 above)
Comment (dblenkus) [reply resolwe/flow/serializers/base.py]: Actually a really bad idea :).

Create a custom error and put it in `flow/exceptions.py`.

File: resolwe/flow/serializers/base.py [hunk 4]
```diff
@@ -36,7 +36,7 @@ class of child class.
 
     contributor = ContributorSerializer()
     name = serializers.CharField(required=False)
-    slug = serializers.CharField(required=False)
+    slug = serializers.CharField(required=False, allow_null=True)  # Allow null to support slug autogeneration
```
Comment (dblenkus) [inline resolwe/flow/serializers/base.py]: Put the comment in a separate line above. And don't forget to put a dot at the end.

File: resolwe/flow/tests/test_api.py [hunk 5]
```diff
@@ -301,6 +301,35 @@ def test_change_descriptor_schema(self):
...
+        response = self.collection_detail_viewset(request, pk=collection1.pk)
+        self.assertEqual(response.data['slug'], 'collection')
+
+        collection3 = Collection.objects.create(name="Collection", contributor=self.contributor)
+        self.assertEqual(collection3.slug, 'collection-3')
+
+        request = factory.patch(self.detail_url(collection2.pk), {'name': 'Collection', 'slug': None}, format='json')
+        force_authenticate(request, self.admin)
+        response = self.collection_detail_viewset(request, pk=collection2.pk)
+        self.assertEqual(response.data['slug'], 'collection-4')
```
Comment (dblenkus) [inline resolwe/flow/tests/test_api.py]: This is probably not the desired behaviour and should be fixed at some point, so i think it is not a good idea to include it in tests.

Comment (hadalin) [reply resolwe/flow/tests/test_api.py]: Is it the problem that the new slug should be `collection-2`?

Comment (hadalin) [reply resolwe/flow/tests/test_api.py]: Fixed in https://github.com/genialis/resolwe/pull/527/commits/97a6c69854191f3024e4b8ae6aaa006a34d3ea16

File: resolwe/flow/serializers/base.py (hunk 3 above)
Comment (hadalin) [reply resolwe/flow/serializers/base.py]: Fixed in https://github.com/genialis/resolwe/pull/527/commits/b45d2f9abff4c82ab13a43515d89c99afad5c738

File: resolwe/flow/tests/test_api.py [hunk 6]
```diff
@@ -301,6 +301,30 @@ def test_change_descriptor_schema(self):
...
+        collection2 = Collection.objects.create(name="Collection", contributor=self.contributor)
+        self.assertEqual(collection1.slug, 'collection')
+        self.assertEqual(collection2.slug, 'collection-2')
+
+        request = factory.patch(self.detail_url(collection1.pk), {'name': 'Collection', 'slug': None}, format='json')
+        force_authenticate(request, self.admin)
+        response = self.collection_detail_viewset(request, pk=collection1.pk)
+        self.assertEqual(response.data['slug'], 'collection')
+
+        collection3 = Collection.objects.create(name="Collection", contributor=self.contributor)
```
Comment (dblenkus) [inline resolwe/flow/tests/test_api.py]: Now you can remove this.

File: resolwe/flow/tests/test_api.py [hunk 7]
```diff
@@ -301,6 +301,30 @@ def test_change_descriptor_schema(self):
...
+        self.assertEqual(collection3.slug, 'collection-3')
+
+        request = factory.patch(self.detail_url(collection2.pk), {'slug': 'collection-2'}, format='json')
```
Comment (dblenkus) [inline resolwe/flow/tests/test_api.py]: This should probably be `collection-3`.

File: resolwe/flow/serializers/base.py [hunk 8]
```diff
@@ -61,9 +62,9 @@ def request(self):
...
         return self.context.get('request', None)
 
-    def save(self):
+    def save(self, **kwargs):
         """Override save method to catch handled errors and repackage them as 400 errors."""
         try:
-            return super(ResolweBaseSerializer, self).save()
-        except DatabaseError as database_error:
-            raise ParseError(database_error)
+            return super(ResolweBaseSerializer, self).save(**kwargs)
```
Comment (dblenkus) [inline resolwe/flow/serializers/base.py]: `ResolweBaseSerializer, self` part is no longer needed in Pyhton 3.

File: resolwe/flow/tests/test_api.py (hunk 5 above)
Comment (dblenkus) [reply resolwe/flow/tests/test_api.py]: It is not consistent with the first use-case in this test, where slugs stays the same. We should at least unify the behaviour.

**Pull request merged**